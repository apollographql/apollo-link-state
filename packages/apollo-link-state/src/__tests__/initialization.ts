import gql from 'graphql-tag';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';

import { withClientState } from '../';

describe('initialization', () => {
  const resolvers = { Query: { foo: () => ({ bar: true }) } };
  const defaults = { foo: { bar: false, __typename: 'Bar' } };

  it('attaches writeData to the cache if you pass in defaults', () => {
    const cache = {
      writeQuery: jest.fn(),
    };

    withClientState({ cache, resolvers, defaults });
    expect('cache.writeData').toBeDefined();
  });

  it('writes defaults to the cache upon initialization', () => {
    const cache = new InMemoryCache();

    withClientState({ cache, resolvers, defaults });
    expect(cache.extract()).toMatchSnapshot();
  });

  it(`doesn't call the resolver if the data is already in the cache`, () => {
    const fooResolver = jest.fn();
    const resolvers = { Query: { foo: fooResolver } };

    const cache = new InMemoryCache();
    const client = new ApolloClient({
      cache,
      link: withClientState({ cache, resolvers, defaults }),
    });

    const query = gql`
      {
        foo @client {
          bar
        }
      }
    `;

    client
      .query({ query })
      .then(({ data }) => {
        expect(fooResolver).not.toHaveBeenCalled();
        expect(data.foo.bar).toEqual(false);
      })
      .catch(e => console.error(e));
  });
});
