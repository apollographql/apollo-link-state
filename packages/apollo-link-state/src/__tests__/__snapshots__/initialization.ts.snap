// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `initialization adds a schema string in SDL format to the context as definition if typeDefs are passed in 1`
] = `
Array [
  Object {
    "definition": "type Todo {
      id: String
      message: String!
    }

    type Query {
      todo(id: String!): Todo
    }",
    "directives": "directive @client on FIELD",
  },
]
`;

exports[
  `initialization concatenates schema strings if typeDefs are passed in as an array 1`
] = `
Array [
  Object {
    "definition": "type Todo {
      id: String
      message: String!
    }

    type Query {
      todo(id: String!): Todo
    }
type Foo {
        foo: String!
        bar: String
      }",
    "directives": "directive @client on FIELD",
  },
]
`;

exports[`initialization writes defaults to the cache upon initialization 1`] = `
Object {
  "$ROOT_QUERY.foo": Object {
    "__typename": "Bar",
    "bar": false,
  },
  "ROOT_QUERY": Object {
    "foo": Object {
      "generated": true,
      "id": "$ROOT_QUERY.foo",
      "type": "id",
    },
  },
}
`;
